File: Seonguk/프로그래머스/광고 삽입/광고_삽입.py
def toSec(time):
    sec=0
    time=time.split(":")
    sec+=int(time[0])*3600
    sec+=int(time[1])*60
    sec+=int(time[2])
    return sec

def solution(play_time, adv_time, logs):
    
    if play_time == adv_time:
        return "00:00:00"
    
    alltime=[0]*(toSec(play_time)+1)
    adv_time=toSec(adv_time)
    for log in logs:
        start, end=log.split("-")
        start=toSec(start)
        end=toSec(end)
        
        # 누적합 표시
        alltime[start]+=1
        alltime[end]-=1
    
    answer=0
    max_val=0
    for i in range(1, len(alltime)):
        alltime[i]+=alltime[i-1]
    for i in range(1, len(alltime)):
        alltime[i]+=alltime[i-1]
        if i >= adv_time-1: # alltime[i] = i-1초까지의 누적 재생시간
            curAdv=alltime[i]-alltime[i-adv_time] # curAdv=i-adv_time초~i초까지의 총 광고 시간
            startAdv=i-adv_time+1
            if max_val < curAdv:
                max_val=curAdv
                answer=startAdv
    
    # 초를 시간 형식으로 바꿔주기
    hour=str(answer//3600)
    hour='0'+hour if len(hour) == 1 else hour
    minute=str((answer%3600)//60)
    minute='0'+minute if len(minute) == 1 else minute
    second=str(answer%3600%60)
    second='0'+second if len(second) == 1 else second
    
    return hour+':'+minute+':'+second

File: LeeGunHee/프로그래머스/광고 삽입 - Lv3/광고삽입.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int cnt[360000];//

int strtoint(string s) {
    string sec = s.substr(6, 2);
    string min = s.substr(3, 2);
    string hour = s.substr(0, 2);

    int result = stoi(sec) + stoi(min) * 60 + stoi(hour) * 60 * 60;
    return result;
}

// 00:00:00 
string inttostr(int i) { 
    int sec = i % 60;
    i /= 60;
    int min = i % 60;
    i /= 60;
    int hour = i;
    string result = "";
    if (hour < 10)result += "0";
    result += to_string(hour);
    result += ":";
    if (min < 10)result += "0";
    result += to_string(min);
    result += ":";
    if (sec < 10)result += "0";
    result += to_string(sec);

    return result;
}


string solution(string play_time, string adv_time, vector<string> logs) {
    string answer = "";
    for (int i = 0; i < logs.size(); i++) { 
        int start = strtoint(logs[i].substr(0, 8));
        int end = strtoint(logs[i].substr(9));
        for (int j = start; j < end; j++) {
            cnt[j]++;
        }
    }
    int playtime = strtoint(play_time); 
    int advtime = strtoint(adv_time);

    queue<int> q; 
    int index = 0;
    long long tmp = 0;
    long long result = 0; 

    for (int i = 0; i < advtime; i++) {
        tmp += cnt[i];
        q.push(cnt[i]);
    }

    result = tmp;
    for (int i = advtime; i < playtime; i++) {
        tmp -= q.front();
        q.pop();
        tmp += cnt[i];
        q.push(cnt[i]);
        if (result < tmp) {
            result = tmp;
            index = i - advtime + 1;
        }
    }
    answer = inttostr(index);

    return answer;
}